x/wasm/keeper/msg_server.go

```diff
@@ -49,38 +49,43 @@ func (m msgServer) StoreCode(goCtx context.Context, msg *types.MsgStoreCode) (*t
 
 // InstantiateContract instantiate a new contract with classic sequence based address generation
 func (m msgServer) InstantiateContract(goCtx context.Context, msg *types.MsgInstantiateContract) (*types.MsgInstantiateContractResponse, error) {
-	return nil, sdkerrors.Wrapf(sdkerrors.ErrNotSupported, "must use x/wormhole")
-	// if err := msg.ValidateBasic(); err != nil {
-	// 	return nil, err
-	// }
-	// ctx := sdk.UnwrapSDKContext(goCtx)
+	if err := msg.ValidateBasic(); err != nil {
+		return nil, err
+	}
+	ctx := sdk.UnwrapSDKContext(goCtx)
 
-	// senderAddr, err := sdk.AccAddressFromBech32(msg.Sender)
-	// if err != nil {
-	// 	return nil, sdkerrors.Wrap(err, "sender")
-	// }
-	// var adminAddr sdk.AccAddress
-	// if msg.Admin != "" {
-	// 	if adminAddr, err = sdk.AccAddressFromBech32(msg.Admin); err != nil {
-	// 		return nil, sdkerrors.Wrap(err, "admin")
-	// 	}
-	// }
+	senderAddr, err := sdk.AccAddressFromBech32(msg.Sender)
+	if err != nil {
+		return nil, sdkerrors.Wrap(err, "sender")
+	}
 
-	// ctx.EventManager().EmitEvent(sdk.NewEvent(
-	// 	sdk.EventTypeMessage,
-	// 	sdk.NewAttribute(sdk.AttributeKeyModule, types.ModuleName),
-	// 	sdk.NewAttribute(sdk.AttributeKeySender, msg.Sender),
-	// ))
+	// check the x/wormhole instantiate allowlist upfront
+	if !m.keeper.HasInstantiateAllowlist(ctx, msg.CodeID, senderAddr) {
```

Comment:
This function is ultimately in the cosmwasm smart contract, right?

Reply:
What do you mean?
This function is called from the token bridge smart contract - that's why we're re-enabling it.
We implemented the allowlist as well since we don't want just anyone to be able to call the wasm instantiate function without a VAA.
The allowlist is maintained in the wormchain `x/wormhole` module.